### PR TITLE
bpo-20392: Fix inconsistency with uppercase file extensions in mimetypes.guess_type

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue20392

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2021-12-22-12-02-27.bpo-20392.CLAFIp.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix inconsistency with uppercase file extensions in :meth:`MimeTypes.guess_type`. Patch by Kumar Aditya.
```
Comment (sobolevn) [inline Misc/NEWS.d/next/Library/2021-12-22-12-02-27.bpo-20392.CLAFIp.rst]: I think it is safe to say that the original patch was authored by @rodriggo.parra

Link: https://bugs.python.org/file33703/case_guess_type.patch

**Review by sobolevn — COMMENTED** [review]
Thanks!

From looking on the original patch, it looks like that you've missed this part:

```diff
-        The suffixes .tgz, .taz and .tz (case sensitive!) are all
+        The suffixes .tgz, .taz and .tz are all
```

Is it intentional?

File: Misc/NEWS.d/next/Library/2021-12-22-12-02-27.bpo-20392.CLAFIp.rst (hunk 1 above)
Comment (kumaraditya303) [reply Misc/NEWS.d/next/Library/2021-12-22-12-02-27.bpo-20392.CLAFIp.rst]: I didn't even read the patch as it was 7 years old and outdated.

**Comment by kumaraditya303** [comment]
> Thanks!
> 
> From looking on the original patch, it looks like that you've missed this part:
> 
> ```diff
> -        The suffixes .tgz, .taz and .tz (case sensitive!) are all
> +        The suffixes .tgz, .taz and .tz are all
> ```
> 
> Is it intentional?

I haven't applied that patch and as you commented I even tried to apply it and it is outdated so not worth copy pasting

File: Lib/mimetypes.py [hunk 2]
```diff
@@ -141,25 +141,22 @@ def guess_type(self, url, strict=True):
                 type = 'text/plain'
             return type, None           # never compressed, so encoding is None
         base, ext = posixpath.splitext(url)
-        while ext in self.suffix_map:
-            base, ext = posixpath.splitext(base + self.suffix_map[ext])
+        while (ext_lower := ext.lower()) in self.suffix_map:
+            base, ext = posixpath.splitext(base + self.suffix_map[ext_lower])
         if ext in self.encodings_map:
```
Comment (JelleZijlstra) [inline Lib/mimetypes.py:147]: From the issue it looks like it's intentional that we don't casefold the encodings_map lookup, but it's not obvious to a reader of the code. Could you add a comment explaining why we shouldn't lower here?

Comment (kumaraditya303) [reply Lib/mimetypes.py:147]: Encodings map is case sensitive hence it should not be lowered, I added a comment.

**Review by JelleZijlstra — APPROVED** [review]
Looks right but I'm not comfortable merging this; I'd like to hear from someone more familiar with MIME types first.

**Review by asvetlov — APPROVED** [review]

**Pull request merged**